### PR TITLE
APPSRE-8160: Handle vault keys with whitespace

### PR DESCRIPTION
**Pull request by lisa**
APPSRE-8160: Handle vault keys with whitespace issues.

### CODE REVIEW

**Comment by esron** [comment]
`make format` can get rid of most formatting errors.

File: reconcile/utils/vault.py
Comment (bhushanthakur93) [inline reconcile/utils/vault.py:199]: We should add this exception in `retry` decorator through `no_retry_exceptions` to avoid unnecessary extra calls.

**Comment by lisa** [comment]
> `make format` can get rid of most formatting errors.

make one simple change and the linter goes nuts.

**Review by bhushanthakur93 — CHANGES_REQUESTED** [review]

File: reconcile/utils/vault.py
Comment (lisa) [reply reconcile/utils/vault.py:199]: Done.

**Comment by TGPSKI** [comment]
LGTM

+ suggest `make format` as prereq for git push (saves time)

After re-tracing the read_all -> read_all_with_version -> _read_all_v1 / read_all_v2, I forgot how complex this read chain is. I'm confident you covered every case, however.

File: reconcile/test/test_vault_utils.py
Comment (chassing) [inline reconcile/test/test_vault_utils.py:56]: Please do this import at the top of the file

File: reconcile/test/test_vault_utils.py
Comment (chassing) [inline reconcile/test/test_vault_utils.py:78]: Do you know https://docs.pytest.org/en/7.3.x/how-to/parametrize.html#pytest-mark-parametrize? with `pytest-mark-parametrize` you could reuse your test function.

**Review by chassing — APPROVED** [review]
Just a few nits. LGTM

File: reconcile/utils/vault.py
Comment (lechuk47) [inline reconcile/utils/vault.py:30]:
```suggestion
class SecretKeyFormatError(Exception):
```
Just a NIT. I think it's more clear

**Comment by lisa** [comment]
/hold

will relocate this to a better (perhaps) spot

File: reconcile/utils/vault.py
Comment (lisa) [reply reconcile/utils/vault.py:30]: 👍 will adopt this as part of the code relocation.

File: reconcile/test/test_vault_utils.py
Comment (lisa) [reply reconcile/test/test_vault_utils.py:56]: For what purpose? It will have no effect for the reason stated in the comment directly preceeding line 56.

File: reconcile/openshift_resources_base.py
Comment (chassing) [inline reconcile/openshift_resources_base.py:610]: Please add type hints.

**Review by chassing — COMMENTED** [review]
Why did you move the validation into `openshift_resources_base` instead of having it in 'utils.vault'?

File: reconcile/openshift_resources_base.py
Comment (bhushanthakur93) [inline reconcile/openshift_resources_base.py:610]: A better check would be to abide by the spec. https://kubernetes.io/docs/concepts/configuration/secret/#restriction-names-data 

> The keys of data and stringData must consist of alphanumeric characters, -, _ or .

I also suggest updating the error message so it's more user friendly. The error message should point out vault path and key, in addition to the spec requirements. Often times I see our tenants struggling with the lack of good error messaging, so small things like this definitely adds up.

File: reconcile/test/test_openshift_resources_base.py
Comment (bhushanthakur93) [inline reconcile/test/test_openshift_resources_base.py:497]: I suggest to split up the tests so that they can verify one behavior at a time. In this case, I suggest you create a separate test that expects the error and another one without error.

File: reconcile/openshift_resources_base.py
Comment (lisa) [reply reconcile/openshift_resources_base.py:610]: that's a good point @bhushanthakur93

File: reconcile/test/test_openshift_resources_base.py
Comment (lisa) [reply reconcile/test/test_openshift_resources_base.py]: I've decided not to do that for two reasons. First, I want to avoid future comments similar to https://github.com/app-sre/qontract-reconcile/pull/3967#discussion_r1431064114. Second, I think the test table approach here is a best fit because the tests all follow the same structure and flow.

Comment (bhushanthakur93) [reply reconcile/test/test_openshift_resources_base.py]: I highly suggest reading https://testing.googleblog.com/2014/04/testing-on-toilet-test-behaviors-not.html.

You can still utilize parametrize tests for different inputs, just isolate the two different behavior of the same function in different tests :) 

That said, I leave this up to you if you still want to keep the same test.

**Review by bhushanthakur93 — APPROVED** [review]

File: reconcile/openshift_resources_base.py
Comment (chassing) [inline reconcile/openshift_resources_base.py:620]: Please adapt the error message too.

**Review by chassing — APPROVED** [review]
Just a nit, otherwise LGTM. Thanks for your work

File: reconcile/openshift_resources_base.py
Comment (lisa) [reply reconcile/openshift_resources_base.py:620]: Whoops. Copy/paste error. ;)

Comment (lisa) [reply reconcile/openshift_resources_base.py]: @chassing Have a look now.

Comment (chassing) [reply reconcile/openshift_resources_base.py]: Perfect. Thank you

**Pull request merged**
